docs/model_developers/modules.rst

```diff
@@ -122,3 +122,60 @@ The format for requires in requirements.txt is the folllowing:
  * Or a repository location such as  git+https://github.com/project/python-foo The correct syntax
    to use is then: eggname@git+https://../repository#branch with branch being optional.
 
+
+Working on modules
+==================
+Since modules often have dependencies on other modules, it is common to develop against multiple
+modules (or a project and one or more modules) simultaneously. One might for example need to
+extend a dependency module to add support for some new feature. Because this use case is so common,
+this section will describe how to work on multiple modules simultaneously so that any changes are
+visible the compiler. This procedure is of course applicable for working on a single module as well.
+
+Setting up the dev environment
+------------------------------
+v1 modules
+^^^^^^^^^^
+A v1 module is a module that is not packaged and published to a pip index but is solely distributed
+via git. V1 modules are installed on the fly by the compiler and can be found in the project's
+`libs` directory. Any changes you make to the module source in the `libs` dir will be reflected in
+the next compile. If you need to be able to make changes to multiple modules simultaneously, the
+simplest way to set up your development environment is to create a project that imports the desired
+modules, then run a compile against it. Alternatively, you can manually clone the module repos.
+
+v2 modules
+^^^^^^^^^^
+A v2 module in development form has mostly the same structure as a v1 module. The main difference is
+that it is meant to be published as a Python package. A project or another v2 module then lists all
+required v2 modules as dependencies in its `pyproject.toml`. The compiler does not install v2
```

Comment:
Why not? 

Doesn't the install behavior depend on the project setup: if you give it module sources of type 'git' or if you give it module source of type 'pypi'

(differentiating the install behavior of V1 and V2 is a bit hard, as you don't know up front what it is going to be)

Reply:
I guess this is something we need to discuss further then (which means this writing the documentation first approach is working). My takeaway from the design discussions was that we wanted to mimic the Python behavior in this regard: a module does not install its own dependencies, instead they must be installed when setting up the environment.
- it's simple in concept, which makes it easy to understand and to implement. No need to think about when and where the compiler installs it and whether it can be persisted in the compiler venv. Single install during env setup and that's it.
- the dependencies are essentially Python dependencies, which is a point in favor of this approach: we stay consistent with the Python conventions.
- it requires an extra step of actually installing the module (which pulls in all dependencies), which is a potential negative for this approach.

@arnaudsjs @bartv this is a design discussion which is probably relevant to all of us.

Reply:
I'm not so much interested in installing a module, but in installing a project (we currently have no 'installing a module' per-se)

Reply:
Well, the same goes for a project in my interpretation.

Reply:
I agree that when you install a Python package it installs its dependencies. That would be the case here as well. What a Python project doesn't do is install its dependencies implicitly when you run/import the module. You need to actively install a package to set up the required environment. If you later remove one of its dependencies from the venv, it doesn't get reinstalled automatically except if you run `pip install .` again.

Reply:
I feel like most other discussions on this pull request depend on this one (some more closely than others). I think this is a pretty fundamental building block of the design. Perhaps we could discuss how we want this behavior in some more detail in tommorow's meeting? I feel like there are two major options:
- The compiler can implicitly install v2 modules via pip if they are not installed in the parent env. This behavior is close to the current behavior which makes it easy to understand. More difficult perhaps is the management of the v2 modules: does the compiler install it in its own venv? What then if the developer installs one of the modules in editable form in the parent venv at some later stage?
- The compiler expects all dependencies to be preinstalled (running `inmanta project install` or `inmanta modules install` could be sufficient). This puts the responsibility for module installation with the developer, allowing them to be more flexible and in control of which dependencies (packaged vs editable, exact version, ...) are installed. Additionally, I feel like this is the more natural approach if we label modules as Python packages, though I must admit the differences with standard Python packages are sufficient that this last argument might not carry that much weight.

Do you agree it would be a good idea to discuss this in a meeting in the hopes of reaching a consensus there before proceeding with the design details that follow from it?

Reply:
(I've context switched to the machine learning project mostly so this is not blocking me in the meantime. I don't think it is blocking the modules v2 in general either for now.)

Reply:
Indeed. There are some loose ends here. I am in favor of a meeting.